tests/Feature/Platform/Action/ResolveAchievementSetsActionTest.php

```diff
@@ -0,0 +1,506 @@
+                ->whereType(AchievementSetType::Bonus)
+                ->first()
+                ->id,
+            'opted_in' => false,
+        ]);
+
+        // Act
+        $resolved = (new ResolveAchievementSetsAction())->execute($specialtyGameHash, $user);
+
+        // Assert
+        $this->assertCount(2, $resolved);
+
+        $this->assertContainsAchievementSetType($resolved, AchievementSetType::Core);
+        $this->assertContainsAchievementSetType($resolved, AchievementSetType::Specialty);
+        $this->assertNotContainsAchievementSetType($resolved, AchievementSetType::Bonus);
+    }
+
+    public function testItReturnsCoreSetsForBonusHashes(): void
+    {
+        // Arrange
+        $baseGame = $this->createGameWithAchievements($this->system, 'Dragon Quest III', 6, 1);
+        $bonusGame = $this->createGameWithAchievements($this->system, 'Dragon Quest III [Subset - Bonus]', 8, 1);
+
+        $this->upsertGameCoreSetAction->execute($baseGame);
+        $this->upsertGameCoreSetAction->execute($bonusGame);
+
+        $this->associateAchievementSetToGameAction->execute($baseGame, $bonusGame, AchievementSetType::Bonus, 'Bonus');
+
+        $bonusGameHash = GameHash::factory()->create(['game_id' => $bonusGame->id]);
+
+        $user = User::factory()->create(['websitePrefs' => self::OPT_IN_TO_ALL_SUBSETS_PREF_ENABLED]);
+
+        // Act
+        $resolved = (new ResolveAchievementSetsAction())->execute($bonusGameHash, $user);
+
+        // Assert
+        $this->assertCount(2, $resolved);
+
+        $coreSet = $resolved->first();
+        $this->assertAchievementSet($coreSet, AchievementSetType::Core, $baseGame->id, 6, 1);
+
+        $bonusSet = $resolved->last();
+        $this->assertAchievementSet($bonusSet, AchievementSetType::Bonus, $baseGame->id, 8, 1);
+    }
+
+    public function testItExcludesSubsetsIfUserIsGloballyOptedOut(): void
+    {
+        // Arrange
+        $baseGame = $this->createGameWithAchievements($this->system, 'Dragon Quest III', 6, 1);
+        $bonusGame = $this->createGameWithAchievements($this->system, 'Dragon Quest III [Subset - Bonus]', 8, 1);
+
+        $this->upsertGameCoreSetAction->execute($baseGame);
+        $this->upsertGameCoreSetAction->execute($bonusGame);
+
+        $this->associateAchievementSetToGameAction->execute($baseGame, $bonusGame, AchievementSetType::Bonus, 'Bonus');
+
+        $baseGameHash = GameHash::factory()->create(['game_id' => $baseGame->id]);
+
+        $user = User::factory()->create(['websitePrefs' => self::OPT_IN_TO_ALL_SUBSETS_PREF_DISABLED]);
+
+        // Act
+        $resolved = (new ResolveAchievementSetsAction())->execute($baseGameHash, $user);
+
+        // Assert
+        $this->assertCount(1, $resolved);
+
+        $coreSet = $resolved->first();
+        $this->assertAchievementSet($coreSet, AchievementSetType::Core, $baseGame->id, 6, 1);
+    }
+
+    public function testItIncludesSubsetIfUserIsGloballyOptedOutButLocallyOptedIn(): void
```

Comment:
I may not have seen them, but what about:
* Globally opted out of subsets, load subset hash -> expect subset and core (or just subset?)
* Explicitly opted out of subset, load subset hash -> expect subset and core (or just core?)

Reply:
These cases were not included. This was really tough to solve, but I think I've landed somewhere I'm happy with.

I've added two new test cases to cover these bullet points. I've also updated some of the behavior of the action. All the existing tests remain untouched & passing.

For your first bullet point:
* The user is **globally opted out** of subsets.
* They load a bonus subset's hash in their emulator.

I think for this case, they should be treated as still living in a pre-multiset environment. The only set that should be resolved is the one for the bonus set, just like what happens today.

For your second bullet point:
* The user is **globally opted in** to subsets.
* The user is opted out of Bonus Set X.
* They load a hash associated with Bonus Set X in their emulator.

For this case, I think the action should behave just like they loaded a hash associated with the base game / core set. They get the core set, any other bonus sets they're not opted out of, and they _don't_ get the subset they are opted out of.

In other words, let's say we have these three sets:
* Dragon Quest III
* Dragon Quest III [Subset - Bonus]
* Dragon Quest III [Subset - Bonus 2]

The user has explicitly opted out of [Subset - Bonus]. They then load a hash associated with [Subset - Bonus]. The core set and [Subset - Bonus 2] are going to be resolved. The action will behave just like they loaded a hash associated with the base game.

This is the only behavior I can think of that keeps things simple and will be easy to document/explain to people. If you're opted out, then the system will respect that no matter what.

A test has been added which verifies the above scenario.